tests/src/Integration/Action/RulesComponentActionTest.php

```diff
@@ -0,0 +1,200 @@
+    $entity = $this->prophesizeEntity(EntityInterface::class);
+    $entity->save()->shouldBeCalledTimes(1);
+
+    RulesComponent::create($rule)
+      ->addContextDefinition('entity', ContextDefinition::create('entity'))
+      ->setContextValue('entity', $entity->reveal())
+      ->execute();
+  }
+
+  /**
+   * Tests that context definitions are available on the derived action.
+   */
+  public function testContextDefinitions() {
+    $rule = $this->rulesExpressionManager->createRule();
+    $rule
+      ->addAction('rules_entity_save', ContextConfig::create()
+        ->map('entity', 'entity')
+      )
+      ->addAction('rules_test_string', ContextConfig::create()
+        ->setValue('text', 'x')
+      );
+
+    $rules_config = new RulesComponentConfig([
+      'id' => 'test_rule',
+      'label' => 'Test rule',
+    ], 'rules_component');
+    $rules_config->setExpression($rule);
+
+    $context_definitions = ['entity' => ContextDefinition::create('entity')];
+    $rules_config->setContextDefinitions($context_definitions);
+    $provided_definitions = ['concatenated' => ContextDefinition::create('string')];
+    $rules_config->setProvidedContextDefinitions($provided_definitions);
+
+    $this->prophesizeStorage([$rules_config]);
+
+    $definition = $this->actionManager->getDefinition('rules_component:test_rule');
+    $this->assertEquals($context_definitions, $definition['context']);
+    $this->assertEquals($provided_definitions, $definition['provides']);
+  }
+
+  /**
+   * Tests that a rules component in an action can also provide variables.
+   */
+  public function testExecutionProvidedVariables() {
+    // Create a rule that produces a provided string variable.
+    $nested_rule = $this->rulesExpressionManager->createRule();
+    $nested_rule->addAction('rules_test_string', ContextConfig::create()
+      ->setValue('text', 'x')
+    );
+
+    $rules_config = new RulesComponentConfig([
+      'id' => 'test_rule',
+      'label' => 'Test rule',
+    ], 'rules_component');
+    $rules_config->setExpression($nested_rule);
+    $rules_config->setProvidedContextDefinitions(['concatenated' => ContextDefinition::create('string')]);
+
+    $this->prophesizeStorage([$rules_config]);
+
+    // Invoke the rules component in another rule.
+    $rule = $this->rulesExpressionManager->createRule();
+    $rule->addAction('rules_component:test_rule');
+
+    $result = RulesComponent::create($rule)
+      ->provideContext('concatenated')
+      ->execute();
+
+    $this->assertEquals('xx', $result['concatenated']);
+  }
+
+  /**
+   * Tests that auto saving is only triggered once with nested components.
+   */
+  public function testAutosaveOnlyOnce() {
+    $entity = $this->prophesizeEntity(EntityInterface::class);
+
+    $nested_rule = $this->rulesExpressionManager->createRule();
+    $nested_rule->addAction('rules_data_set', ContextConfig::create()
+      ->map('data', 'entity')
```

Comment:
I think the test coverage is bogus as setting an "entity" variable with a new entity should NOT trigger auto-saving at all. It should only save the entity if you change it, not if you set some variable with it. It might have to save the parent data structure though, but that's another issue. I think the test case should just use the entity save action.